graylog2-server/src/main/java/org/graylog2/rest/resources/datanodes/DatanodeRestApiProxy.java

```diff
@@ -26,45 +30,99 @@
 import okhttp3.Response;
 import org.apache.commons.io.IOUtils;
 import org.apache.commons.lang3.StringUtils;
+import org.apache.http.HttpStatus;
 import org.graylog2.cluster.nodes.DataNodeDto;
 import org.graylog2.cluster.nodes.NodeDto;
 import org.graylog2.cluster.nodes.NodeService;
 import org.graylog2.indexer.datanode.ProxyRequestAdapter;
 import org.graylog2.security.IndexerJwtAuthTokenProvider;
 import org.jetbrains.annotations.NotNull;
 
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
 import java.io.IOException;
-import java.security.KeyManagementException;
-import java.security.NoSuchAlgorithmException;
-import java.util.Collection;
+import java.io.InputStream;
+import java.net.http.HttpResponse;
+import java.time.Duration;
 import java.util.Locale;
+import java.util.Map;
 import java.util.Objects;
-import java.util.Optional;
+import java.util.stream.Collectors;
 
 @Singleton
 public class DatanodeRestApiProxy implements ProxyRequestAdapter {
 
     private final IndexerJwtAuthTokenProvider authTokenProvider;
+    private final NodeService<DataNodeDto> nodeService;
+    private final ObjectMapper objectMapper;
     private final DatanodeResolver datanodeResolver;
     private final OkHttpClient httpClient;
 
     @Inject
-    public DatanodeRestApiProxy(IndexerJwtAuthTokenProvider authTokenProvider, DatanodeResolver datanodeResolver, OkHttpClient okHttpClient) throws NoSuchAlgorithmException, KeyManagementException {
+    public DatanodeRestApiProxy(IndexerJwtAuthTokenProvider authTokenProvider, NodeService<DataNodeDto> nodeService, ObjectMapper objectMapper, DatanodeResolver datanodeResolver, OkHttpClient okHttpClient, @Named("proxied_requests_default_call_timeout")
+    com.github.joschi.jadconfig.util.Duration defaultProxyTimeout) {
         this.authTokenProvider = authTokenProvider;
+        this.nodeService = nodeService;
+        this.objectMapper = objectMapper;
         this.datanodeResolver = datanodeResolver;
-        httpClient = okHttpClient;
+        httpClient = withTimeout(okHttpClient, defaultProxyTimeout);
+    }
+
+    @NotNull
+    private static OkHttpClient withTimeout(OkHttpClient okHttpClient, com.github.joschi.jadconfig.util.Duration defaultProxyTimeout) {
+        final Duration timeout = Duration.ofMillis(defaultProxyTimeout.toMilliseconds());
+        return okHttpClient.newBuilder()
+                .connectTimeout(timeout)
+                .readTimeout(timeout)
+                .callTimeout(timeout)
+                .build();
+    }
+
+    private ProxyResponse runOnAllNodes(ProxyRequest request) {
+        try (final ByteArrayOutputStream baos = copyRequestBody(request)) {
+            final Map<String, JsonNode> result = nodeService.allActive().values().stream().parallel().collect(Collectors.toMap(NodeDto::getHostname, n -> {
```

Comment:
It is not smth that needs to be addressed now, if the PR is urgent, but I have 2 small ideas:
1. I'd use more new lines for better readability, i.e.:
```
nodeService
.allActive()
.values()
.stream()
.parallel()
```
2. You could consider mapping from `n` to hostname between `parallel()` and `collect()`